src/std/sugar.ss

```diff
@@ -208,3 +227,81 @@
   ((_ (id test) body ...)
    (let (id test)
      (when id body ...))))
+
+;; chain rewrites passed expressions by passing the previous expression
+;; into the possition of the <> diamond symbol. In case a previous expression
+;; should be used in a sub-expression, or multiple times, the expression can be
+;; prefixed with a variable (supports destructuring).
+;;
+;; When the first expression is a <> or ([pattern] <> expression),
+;; chain will return a unary lambda.
+;;
+;; Example:
+;;  (chain [1 2 3]
+;;         ([_ . rest] (map number->string rest))  ; with
+;;         (v (string-join v ", "))                ; let
+;;         (string-append <> " :)"))
+;; => "2, 3 :)"
+(defrules chain (<>)
+  ((_ <> (fn arg arg* ...) ...)
+   (lambda (v)
+     (~chain-aux ((fn arg arg* ...) ...) v)))
+
+  ((_ (vars <> exp) (fn arg arg* ...) ...)
+   (lambda (v)
+     (with ((vars v))
+       (~chain-aux ((fn arg arg* ...) ...) exp))))
+
+  ((_ exp (fn arg arg* ...) ...)
+   (~chain-aux ((fn arg arg* ...) ...) exp)))
+
+;; ~chain-aux is an auxiliary macro which takes a list of expressions
+;; and the initial chain value. It then loops over the expression list
+;; and transforms one expression after the other.
+(defrules ~chain-aux (<>)
+  ((_ () acc)
+   acc)
+
+  ((_ ((var ()) . more) acc)
+   (syntax-error "Body expression cannot be empty"))
+
+  ;; variable
+  ((_ ((var (body1 body2 . body*)) . more) acc)
+   (~chain-aux more
+	      (~chain-aux-variable (var acc) (body1 body2 . body*))))
+
+  ;; diamond
+  ((_ ((fn . args) . more) acc)
+   (~chain-aux more
+	      (~chain-aux-diamond (fn . args) () acc))))
+
+;; ~chain-aux-variable is an auxiliary macro which transforms an expression
+;; either in a with or let expression.
+(defrules ~chain-aux-variable ()
+  ((_ (() (fn . args)) body)
+   (syntax-error "the variable must be non-empty"))
+
+  ((_ ((var . vars) previous) body)
+   (with (((var . vars) previous)) body))
+
+  ((_ (var previous) body)
+   (let ((var previous)) body)))
```

Comment:
this should probably also be a `with`?

Reply:
This is an interesting question. I thought about it too and compared the `@expand` output of `let` and `with` when no destructuring is used. Since the output is different and the one of `let` seems "lighter", I chose `let` (the use should only "pay" – resource wise – for what he uses). Another option would be to use a `lambda` and call it directly.

What is the fastest/best solution here?

Reply:
it's probably the same when compiling.